### ISSUE TITLE
EKSA Bare Metal: cluster ScaleUp operation fails because Kubelet fails to start on new bottlerocket-os node (Additionally sshd and agetty don't start either making debugging impossibly hard)

### ISSUE THREAD

**Issue by robbierafay** [aws/eks-anywhere#5564]
**What happened**: Scale Up operation failed for a 2 node EKSA Bare Metal cluster (1 control-plane node, 1 worker node : both running bottlerocket-os) where the Kubelet failed to start on newly added  bottlerocket-os node . 

**Along with the failure of the Kubelet, additionally  SSH and agetty also didn't start and hence there is no way to login to the bottlerocket node and debug Kubelet. Hence I am stuck!!**

Below is the error on the serial console of the new node:

```
..............
..............
[  OK  ] Reached target Login Prompts.
[  OK  ] Reached target Remote File Systems.
         Starting containerd container runtime...
         Starting Load Kernel Module ip_vs_rr...
         Starting Load Kernel Module ip_vs_sh...
         Starting Load Kernel Module ip_vs_wrr...
         Starting Reboot system if needed fo…guration changes to take effect...
[  224.780468] IPVS: Registered protocols (TCP, UDP, SCTP, AH, ESP)
         Startin[  224.805139] IPVS: Connection hash table configured (size=4096, memory=32Kbytes)
g Send boot success...[  224.829281] IPVS: ipvs loaded.

[  OK  ] Finished Reboot system if needed fo…figuration changes to take effect.
[  OK  ] Finished Isolates multi-user.target.
[  224.901412] IPVS: [rr] scheduler registered.
[  224.914451] IPVS: [sh] scheduler registered.
[  224.932628] IPVS: [wrr] scheduler registered.
[  OK  ] Finished Load Kernel Module ip_vs_rr.
[  OK  ] Finished Load Kernel Module ip_vs_sh.
[  OK  ] Finished Load Kernel Module ip_vs_wrr.
[  OK  ] Started containerd container runtime.
         Starting Kubelet...
[  OK  ] Finished Send boot success.
[FAILED] Failed to start Kubelet.
See 'systemctl status kubelet.service' for details.
[  OK  ] Reached target Multi-User System.
```
There are no more messages on the serial console and no login prompt. Also ssh is not responding to the node. 

 

**What you expected to happen**: The cluster should have successfully scaled up.

**How to reproduce it (as minimally and precisely as possible)**:
Below are steps I followed:

1.  Created cluster 
[cluster.yaml.txt](https://github.com/aws/eks-anywhere/files/11189681/cluster.yaml.txt)

$ eksctl anywhere version
v0.15.1

$ eksctl anywhere create cluster --force-cleanup -f ./cluster-rgillnative.yaml -z ./hardware.csv
Warning: The recommended number of control plane nodes is 3 or 5
Warning: The recommended number of control plane nodes is 3 or 5
Performing setup and validations
✅ Tinkerbell Provider setup is valid
✅ Validate OS is compatible with registry mirror configuration
✅ Validate certificate for registry mirror
✅ Validate authentication for git provider
Creating new bootstrap cluster
Provider specific pre-capi-install-setup on bootstrap cluster
Installing cluster-api providers on bootstrap cluster
Provider specific post-setup
Creating new workload cluster
Installing networking on workload cluster
Creating EKS-A namespace
Installing cluster-api providers on workload cluster
Installing EKS-A secrets on workload cluster
Installing resources on management cluster
Moving cluster management from bootstrap to workload cluster
Installing EKS-A custom components (CRD and controller) on workload cluster
Installing EKS-D components on workload cluster
Creating EKS-A CRDs instances on workload cluster
Installing GitOps Toolkit on workload cluster
GitOps field not specified, bootstrap flux skipped
Writing cluster config file
Deleting bootstrap cluster
🎉 Cluster created!

2.  Updated cluster config to increase node count from 1 to 2 in workerNodeGroupConfiguration  and created new hardware.csv

```
$ diff -u ./rgillnative/rgillnative-eks-a-cluster.yaml-ORIG ./rgillnative/rgillnative-eks-a-cluster.yaml
--- ./rgillnative/rgillnative-eks-a-cluster.yaml-ORIG	2023-04-09 07:49:59.020065882 +0000
+++ ./rgillnative/rgillnative-eks-a-cluster.yaml	2023-04-09 07:54:47.859573869 +0000
@@ -32,7 +32,7 @@
   managementCluster:
     name: rgillnative
   workerNodeGroupConfigurations:
-  - count: 1
+  - count: 2
     machineGroupRef:
       kind: TinkerbellMachineConfig
       name: dp-mc

```

$ cat hardware-exp.csv
hostname,mac,ip_address,netmask,gateway,nameservers,labels,disk,bmc_ip,bmc_username,bmc_password
rgillnative-dp-n-2,08:00:27:42:C9:5B,192.168.10.188,255.255.255.0,192.168.10.1,8.8.8.8,type=dp,/dev/sda,,,

3.  Perform scaleUp operation by executing below command 

$ eksctl anywhere upgrade cluster --force-cleanup -f ./rgillnative/rgillnative-eks-a-cluster.yaml -z ./hardware-exp.csv
Warning: The recommended number of control plane nodes is 3 or 5
Warning: The recommended number of control plane nodes is 3 or 5
Warning: The recommended number of control plane nodes is 3 or 5
Performing setup and validations
✅ Tinkerbell provider validation
✅ Validate OS is compatible with registry mirror configuration
✅ Validate certificate for registry mirror
✅ Control plane ready
✅ Worker nodes ready
✅ Nodes ready
✅ Cluster CRDs ready
✅ Cluster object present on workload cluster
✅ Upgrade cluster kubernetes version increment
✅ Validate authentication for git provider
✅ Validate immutable fields
Ensuring etcd CAPI providers exist on management cluster before upgrade
Pausing EKS-A cluster controller reconcile
Pausing GitOps cluster resources reconcile
Upgrading core components
Creating bootstrap cluster
Provider specific pre-capi-install-setup on bootstrap cluster
Provider specific post-setup
Installing cluster-api providers on bootstrap cluster
Backing up workload cluster's management resources before moving to bootstrap cluster
Moving cluster management from workload to bootstrap cluster
Upgrading workload cluster
collecting cluster diagnostics
collecting management cluster diagnostics
⏳ Collecting support bundle from cluster, this can take a while	{"cluster": "bootstrap-cluster", "bundle": "rgillnative/generated/bootstrap-cluster-2023-04-09T08:30:27Z-bundle.yaml", "since": 1681018227898214539, "kubeconfig": "rgillnative/generated/rgillnative.kind.kubeconfig"}
Support bundle archive created	{"path": "support-bundle-2023-04-09T08_30_29.tar.gz"}
Analyzing support bundle	{"bundle": "rgillnative/generated/bootstrap-cluster-2023-04-09T08:30:27Z-bundle.yaml", "archive": "support-bundle-2023-04-09T08_30_29.tar.gz"}
Analysis output generated	{"path": "rgillnative/generated/bootstrap-cluster-2023-04-09T08:30:43Z-analysis.yaml"}
collecting workload cluster diagnostics
⏳ Collecting support bundle from cluster, this can take a while	{"cluster": "rgillnative", "bundle": "rgillnative/generated/rgillnative-2023-04-09T08:30:50Z-bundle.yaml", "since": 1681018250316128996, "kubeconfig": "rgillnative/rgillnative-eks-a-cluster.kubeconfig"}
Support bundle archive created	{"path": "support-bundle-2023-04-09T08_30_53.tar.gz"}
Analyzing support bundle	{"bundle": "rgillnative/generated/rgillnative-2023-04-09T08:30:50Z-bundle.yaml", "archive": "support-bundle-2023-04-09T08_30_53.tar.gz"}
Analysis output generated	{"path": "rgillnative/generated/rgillnative-2023-04-09T08:31:15Z-analysis.yaml"}
Error: failed to upgrade cluster: waiting for workload cluster machinedeployment replicas to be ready: retries exhausted waiting for machinedeployment replicas to be ready: machine deployment is in ScalingUp phase
[support-bundle-2023-04-09T08_30_29.tar.gz](https://github.com/aws/eks-anywhere/files/11189713/support-bundle-2023-04-09T08_30_29.tar.gz)
[support-bundle-2023-04-09T08_30_53.tar.gz](https://github.com/aws/eks-anywhere/files/11189714/support-bundle-2023-04-09T08_30_53.tar.gz)

**Debug Data Collected** 

From KIND cluster:

$ docker exec -ti rgillnative-eks-a-cluster-control-plane kubectl get machines.cluster.x-k8s.io -A
NAMESPACE     NAME                                CLUSTER       NODENAME                            PROVIDERID                                    PHASE         AGE    VERSION
eksa-system   rgillnative-fhlh9                   rgillnative   rgillnative-fhlh9                   tinkerbell://eksa-system/rgillnative-cp-n-1   Running       112m   v1.25.8-eks-1-25-10
eksa-system   rgillnative-md-0-65b745fd88-7hdnv   rgillnative                                       tinkerbell://eksa-system/rgillnative-dp-n-2   Provisioned   110m   v1.25.8-eks-1-25-10
eksa-system   rgillnative-md-0-65b745fd88-sw5zg   rgillnative   rgillnative-md-0-65b745fd88-sw5zg   tinkerbell://eksa-system/rgillnative-dp-n-1   Running       112m   v1.25.8-eks-1-25-10

$ docker exec -ti rgillnative-eks-a-cluster-control-plane kubectl get workflows -A
NAMESPACE     NAME                                   TEMPLATE                               STATE
eksa-system   rgillnative-md-0-1681024626620-flwrz   rgillnative-md-0-1681024626620-flwrz   STATE_SUCCESS

Detailed debug info: [debug-kind-cluster.txt](https://github.com/aws/eks-anywhere/files/11189878/debug-kind-cluster.txt)

**Anything else we need to know?**:

**Environment**:
- EKS Anywhere Release: v0.15.1

- EKS Distro Release:

**Comment by raykrueger** [comment]
Robbie, is this on VirtualBox or bare metal?

**Comment by robbierafay** [comment]
@raykrueger I see it on both

**Comment by robbierafay** [comment]
I decided to mount the failed bottlerocket node's disk on another ubuntu20 system. This way, I was able to get access to the bottlerocket private partition to check configuration and also inspect kubelet logs. 

The issue seems like user-data.toml is not populated correctly on the problem node.  Below are contents of this file on problem node (Note all Kubernetes config is missing)

```
# cat user-data.toml

[settings.dns]
name-servers = ["8.8.8.8"]
```

Checking Kubelet logs, it seems to be failing because CA file /etc/kubernetes/pki/ca.crt does not exist

```Apr 26 14:44:41 192.168.10.226 kubelet[26076]: E0426 14:44:41.278381   26076 server.go:271] "Failed to construct kubelet dependencies" err="unable to load client CA file /etc/kubernetes/pki/ca.crt: error creating pool from /etc/kubernetes/pki/ca.crt: data does not contain any valid RSA or ECDSA certificates"```

Does this mean issue is in Hegel service somewhere ??

**Comment by robbierafay** [comment]
Finally root caused the issue. Seem like the problem was being caused by HEGEL_URLS settings in TinkerbellTemplateConfig. 

Problem setting was 

`HEGEL_URLS: http://tinberbell_ip:50061/,http://admin_machine_ip:50061/`

Changing it to below fixed the issue  

`HEGEL_URLS: http://tinberbell_ip:50061/,http://admin_machine_ip:50061/`

Is this a race condition when HEGEL service is operational on both the workload cluster being scaled and the bootstrap/kind cluster being created by the eksctl anywhere upgrade cluster command ?

**Comment by chrisdoherty4** [comment]
Hi @robbierafay. We've tried to reproduce the issue without success. Is this still a problem? Have you tried the latest release (v0.16) and observed the problem still exists?

**Comment by robbierafay** [comment]
@chrisdoherty4 I was able to solve the issue by switching the order of HEGEL_URLS in TinkerbellTemplateConfig and listing the admin node's IP first:

Problem order:

HEGEL_URLS: http://tinberbell_ip:50061/,http://admin_machine_ip:50061/


Solution: 

HEGEL_URLS: http://admin_machine_ip:50061/,http://tinberbell_ip:50061/


I have not tested if the problem order is still an issue in v0.16. I think the issue might be caused because HEGEL pod in cluster is also responding at the same time when bootstrap (kind) cluster's HEGEL pod is operational. Perhaps we can disable HEGEL service in cluster when bootstrap cluster is operational ? (for instance during ekctl anywhere upgrade)

**Comment by chrisdoherty4** [comment]
@robbierafay When you first created this cluster, did you use a `TinkerbellTemplateConfig` to override some actions?

**Comment by robbierafay** [comment]
@chrisdoherty4 yes I had to use a custom template as default generated template did not work with my node's interface name.  

https://github.com/aws/eks-anywhere/files/11189681/cluster.yaml.txt

**Comment by jiayiwang7** [comment]
@robbierafay the fix will be included in the upcoming v0.17 release. Thanks for the discovery!

**Comment by robbierafay** [comment]
Thanks for the fix @chrisdoherty4 and @jiayiwang7
